Userland/Libraries/LibWeb/Infra/Numbers.cpp

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2023, Jonatan Klemets <jonatan.r.klemets@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/GenericLexer.h>
+#include <LibWeb/Infra/CharacterTypes.h>
+#include <LibWeb/Infra/Numbers.h>
+
+namespace Web::Infra {
+
+// This implements both the signed and unsigned parsing.
+// https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#rules-for-parsing-integers
+// https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#rules-for-parsing-non-negative-integers
+template<typename T>
+Optional<T> to_integer(StringView string)
```

Comment:
Please split this into one function for each algorithm. You can limit each template to only accept unsigned/signed integer types using `requires`.

Though, we may not need to return all the different integer sizes, i32/u32 is probably fine.

Also, as a nit-pick, consider renaming them `parse_integer()` and `parse_non_negative_integer()` so that they match the names of the spec algorithms a little more closely.

Reply:
I opted to just keep the i32/u32 version and also renamed them.